File: challenge/solution/providers.py
"""
Mock provider adapter.

Loads enrichment_responses.json and exposes one method per provider.
Each method returns a typed dict or None if the provider has no data
for that company (intentional "not found" — treat as absence, not error).
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import Optional, TypedDict

_MOCK_PATH = Path(__file__).parent.parent / "mocks" / "enrichment_responses.json"


class RegistryResult(TypedDict):
    name: Optional[str]
    role: Optional[str]
    source_url: str


class ListingResult(TypedDict):
    name: Optional[str]
    phone: Optional[str]
    source_url: str


class EnrichmentResult(TypedDict):
    email: Optional[str]
    phone: Optional[str]
    provider_confidence: int
    source_url: str


class ProviderResponse(TypedDict):
    registry: Optional[RegistryResult]
    listing: Optional[ListingResult]
    enrichment: Optional[EnrichmentResult]


def _load() -> dict:
    with _MOCK_PATH.open() as f:
        return json.load(f)


_DATA: dict = _load()


def query_all(company_name: str) -> ProviderResponse:
    """Return all three providers' responses for a company, or None per provider if absent."""
    record = _DATA.get(company_name, {})
    return ProviderResponse(
        registry=record.get("registry"),
        listing=record.get("listing"),
        enrichment=record.get("enrichment"),
    )

File: solution/src/providers.js
/**
 * providers.js
 *
 * Loads the canned mock-provider data from enrichment_responses.json
 * and exposes a per-company lookup that returns { registry, listing, enrichment }
 * (any key may be missing – that is a "not found" from that source).
 */

import { readFileSync } from "node:fs";
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";

const __dirname = dirname(fileURLToPath(import.meta.url));

const MOCK_PATH = resolve(
  __dirname,
  "../../challenge/mocks/enrichment_responses.json"
);

let _cache = null;

/** Load once, cache in memory */
function loadMocks() {
  if (!_cache) {
    _cache = JSON.parse(readFileSync(MOCK_PATH, "utf-8"));
  }
  return _cache;
}

/**
 * Query all three mock providers for a given company.
 *
 * @param {string} companyName – exact key from companies.csv
 * @returns {{ registry?: object, listing?: object, enrichment?: object }}
 *          Missing keys mean the provider returned nothing.
 */
export function queryProviders(companyName) {
  const mocks = loadMocks();
  return mocks[companyName] ?? {};
}
